cmd/device-register/main.go

```diff
@@ -0,0 +1,181 @@
+	deviceConfigFile        = "/etc/cacophony/device.yaml"
+	devicePrivateConfigFile = "/etc/cacophony/device-priv.yaml"
+	minionIDPrefix          = "pi-"
+)
+
+var version = "<not set>"
+
+type Args struct {
+	Reboot             bool   `arg:"-r,--reboot" help:"reboot device after registering"`
+	API                string `arg:"-a,--api" help:"url for the api server to register to"`
+	IgnoreMinionID     bool   `arg:"-i,--ignore-minion-id" help:"don't check or write to minion id file"`
+	RemoveDeviceConfig bool   `arg:"-d,--remove-device-Config" help:"remove the device config files. This is useful if you need to register as a new device or to a different server. This normally wants to be used with '-i'"`
+}
+
+func (Args) Version() string {
+	return version
+}
+
+func procArgs() Args {
+	args := Args{
+		API: apiURL,
+	}
+	arg.MustParse(&args)
+	return args
+}
+
+func main() {
+	err := runMain()
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+func runMain() error {
+	log.SetFlags(0) // Removes default timestamp flag
+	args := procArgs()
+	log.Printf("running version: %s", version)
+
+	apiURL, err := url.ParseRequestURI(args.API)
+	if err != nil {
+		return err
+	}
+	apiString := apiURL.String()
+
+	if !args.IgnoreMinionID {
+		if err := checkMinionIDFile(); err != nil {
+			return err
+		}
+	}
+
+	cr := connrequester.NewConnectionRequester()
+	log.Println("requesting internet connection")
+	cr.Start()
+	cr.WaitUntilUpLoop(connectionTimeout, connectionRetryInterval, -1)
+	log.Println("internet connection made")
+
+	if args.RemoveDeviceConfig {
+		if err := deleteDeviceConfigFiles(); err != nil {
+			return err
+		}
+	}
+
+	rand.Seed(time.Now().UnixNano())
+	deviceName := petname.Generate(3, "-")
+	apiClient, err := api.Register(deviceName, randString(20), defaultGroup, apiString)
+	if err != nil {
+		return err
+	}
+	cr.Stop()
+	log.Println("registered")
+	log.Printf("devicename: '%s', deviceID: '%d', API: '%s'", deviceName, apiClient.DeviceID(), apiString)
+
+	if !args.IgnoreMinionID {
+		if err := writeToMinionIDFile(apiClient.DeviceID()); err != nil {
+			return err
+		}
+	}
+
+	if args.Reboot {
+		log.Println("restarting device")
+		if err := exec.Command("reboot").Run(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func writeToMinionIDFile(deviceID int) error {
+	f, err := os.Create(minionIDFile)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	name := minionIDPrefix + strconv.Itoa(deviceID)
+	log.Printf("setting minion id to '%s'", name)
+	if _, err := f.WriteString(name); err != nil {
+		return err
+	}
+	return nil
+}
+
+func deleteDeviceConfigFiles() error {
+	if err := removeFileIfExist(deviceConfigFile); err != nil {
+		return err
+	}
+	if err := removeFileIfExist(devicePrivateConfigFile); err != nil {
+		return err
+	}
+	return nil
+}
+
+func removeFileIfExist(path string) error {
+	if _, err := os.Stat(path); err == nil {
+		if err := os.Remove(path); err != nil {
+			return err
+		}
+	}
```

Comment:
Instead of "looking before you leap" the more efficient and non-racy approach here is to just try and remove the file and then check the error:
```go
err := os.Remove(path)
if err == nil || os.IsNotExist(err) {
    return nil
}
return err
```